File: c_leetcode/maxLengthBetweenEqualCharacters.c
#include "stdio.h"
#include <assert.h>
#include "string.h"

#define MAX(x,y) (x > y ? x : y)
int maxLengthBetweenEqualCharacters(char* s)
{
    int ans = -1;
    int map[26];
    int length = strlen(s);

    memset(map, -1, sizeof(map));
    for (int i = 0; i < length; i++) {
        if (map[s[i] - 'a'] < 0) {
            map[s[i] - 'a'] = i;
        } else {
            ans = MAX(ans, i - map[s[i] - 'a'] - 1);
        }
    }
    return ans;
}

int main()
{
    assert(maxLengthBetweenEqualCharacters("aa") == 0);
    assert(maxLengthBetweenEqualCharacters("abca") == 2);
    assert(maxLengthBetweenEqualCharacters("cbzxy") == -1);
    assert(maxLengthBetweenEqualCharacters("cabbac") == 4);
    return 0;
}

File: rust_leetcode/src/bin/max_length_between_equal_characters.rs
struct Solution;

impl Solution {
    pub fn max_length_between_equal_characters(s: String) -> i32 {
        let mut ans = -1;
        let mut map = [-1; 26];

        for (i, c) in s.bytes().enumerate() {
            if map[(c - b'a') as usize] < 0 {
                map[(c - b'a') as usize] = i as i32;
            } else {
                ans = ans.max(i as i32 - map[(c - b'a') as usize] - 1);
            }
        }

        ans
    }
}

fn main() {
    let s = "scayofdzca".to_string();

    let ans = Solution::max_length_between_equal_characters(s);

    println!("{ans}");
}

#[cfg(test)]
mod tests {
    use crate::Solution;
    #[test]
    fn test() {
        assert_eq!(
            0,
            Solution::max_length_between_equal_characters("aa".to_string())
        );
        assert_eq!(
            2,
            Solution::max_length_between_equal_characters("abca".to_string())
        );
        assert_eq!(
            -1,
            Solution::max_length_between_equal_characters("cbzxy".to_string())
        );
        assert_eq!(
            4,
            Solution::max_length_between_equal_characters("cabbac".to_string())
        );
    }
}
